recipes/lora_finetune_distributed.py

```diff
@@ -270,6 +278,122 @@ def setup(self, cfg: DictConfig) -> None:
+
+        Args:
+            cfg_profiler: DictConfig - `profiler` section of the top-level `cfg` (the main config passed to `recipe.main`)
+            log_cfg: bool - whether to return the profiler config after profiler setup, which sets defaults and possibly
+            overrides certain profiling options.
+
+            NOTE: Since not all settings of the profiler can be parsed from the returned profiler object,
+            such as the `schedule`, `log_cfg` can be used for easy logging / debugging of all profiler options post setup.
+
+        Returns:
+            profiler: torch.profiler.profile | FakeProfiler - FakeProfiler is a nullcontext with no-op methods
+            for `start`, `stop`, and `step` that can be used in place of `torch.profiler.profile` if profiler is not enabled such
+            that the instrumented training loop does not need to be changed profiling is disabled.
+            profiler_cfg: Optional[DictConfig]
+
+        The profiler config can be provided in configs under the `profiler` key with the following layout:
+        ```
+        profiler:
+            enabled: bool
+
+            #Output directory of trace artifacts
+            output_dir: str
+
+            #`torch.profiler.ProfilerActivity` types to trace
+            CPU: bool
+            CUDA: bool
+
+            #Trace options
+            profile_memory: bool
+            with_stack: bool
+            record_shapes: bool
+            with_flops: bool
+
+            #`torch.profiler.schedule` args
+            schedule:
+                wait: int
+                warmup: int
+                active: int
+                repeat: int
+        ```
+        """
+
+        enabled = cfg_profiler is not None and cfg_profiler.get("enabled", True)
+
+        if not enabled:
+            if self._is_rank_zero:
+                log.info(" Profiling disabled.")
+            return FakeProfiler()
+
+        # Set up profiler activities
+        cpu = cfg_profiler.get("CPU", True)
+        cuda = cfg_profiler.get("CUDA", True)
+
+        profile_memory = cfg_profiler.get(
+            "profile_memory", DEFAULT_TRACE_OPTS["profile_memory"]
+        )
+        with_stack = cfg_profiler.get("with_stack", DEFAULT_TRACE_OPTS["with_stack"])
+        record_shapes = cfg_profiler.get(
+            "record_shapes", DEFAULT_TRACE_OPTS["record_shapes"]
+        )
+        with_flops = cfg_profiler.get("with_flops", DEFAULT_TRACE_OPTS["with_flops"])
+        output_dir = cfg_profiler.get("output_dir", DEFAULT_PROFILE_DIR)
+
+        # Parse schedule specific args
+        schedule_cfg = cfg_profiler.get("schedule", None)
+
+        # Check for schedule
+        # 1) If no schedule is provided, set to DEFAULT_SCHEDULE
+        # 2) else check for missing keys and warn if any are missing, setting these to defaults
+        if schedule_cfg is None:
+            schedule_cfg = DEFAULT_SCHEDULE
+        else:
+            schedule_cfg = {
```

Comment:
nit: I think this can be simplified further:
```
for k in DEFAULT_SCHEDULE.keys():
    if k not in schedule_cfg:
        schedule_cfg[k] = DEFAULT_SCHEDULE[k]
        log.warning(f"Missing key {k} in schedule config, defaulting to {k} = {schedule_cfg[k]}")
```